### DISCUSSION TITLE
Next 13 adds a new tsconfig.json field — "plugins": [{ "name": "next" }]. What is its purpose?

### DISCUSSION THREAD

**Discussion by dvakatsiienko** [vercel/next.js#42285] [Help] [answered]
And if there are any docs for it?

```json
{
    "include": ["next-env.d.ts", "./src", "./types", "./mocks"],
    "exclude": ["node_modules", "cypress"],
    "ts-node": { "compilerOptions": { "module": "commonjs" } },
    "compilerOptions": {
        // ...rest of the config

        /* Completeness */
        "skipLibCheck": true,
        "incremental": true,
        "plugins": [{ "name": "next" }] // this option?
    }
}
```

**Comment by icyJoseph** [comment, marked as answer]
These are ways for the plugin to "share more typing validation" with the client (you). For example an SQL plugin would read type errors inside templated strings.

It lives here, as of now, https://github.com/vercel/next.js/blob/canary/packages/next/server/next-typescript.ts

It can do things such as forbid showing `useState` as an import from `react` when the 'use client' directive is not present.

And in the docs you find it here: https://beta.nextjs.org/docs/configuring/typescript#editor-configuration

**Comment by dvakatsiienko** [comment]
@icyJoseph cool. thanks.
